`test foo =~ foo` should fail with exit status 2

JohnoKing opened the pull request:
When `test` is passed the '=~' operator, it will silently fail with exit status 1:  
```sh
$ test foo =~ foo; echo $?
1
```
This bug is caused by `test_binop` reaching the `NOTREACHED` section of code. The bugfix was backported from ksh2020: https://github.com/att/ast/issues/1152

src/cmd/ksh93/bltins/test.c: test_binop():
- Error out with a message suggesting usage of `[[...]]` if '=~' is passed to the `test` builtin.
- Add an `abort()` under a default case label. This section of code should never be reached when test is given an unknown operator.

src/cmd/ksh93/tests/builtins.sh:
- Add regression tests for how `test` handles the invalid '=~' and '===' operators.

JohnoKing committed:
`test foo =~ foo` should fail with exit status 2

When test is passed the '=~' operator, it will silently fail with
exit status 1:
    $ test foo =~ foo; echo $?
    1
This bug is caused by test_binop reaching the 'NOTREACHED' area of
code. The bugfix was backported from ksh2020:
https://github.com/att/ast/issues/1152

src/cmd/ksh93/bltins/test.c: test_binop():
- Error out with a message suggesting usage of '[[...]]' if '=~' is
  passed to the test builtin.
- Add an abort() under a default case label. This section of code
  should never be reached when test is given an unknown operator.

src/cmd/ksh93/tests/builtins.sh:
- Add regression tests for how 'test' handles the invalid '=~'
  and '===' operators.

hyenias commented on src/cmd/ksh93/tests/builtins.sh:
```diff
@@ -996,6 +996,28 @@ then	got=$( { "$SHELL" -c '
 
+# ==========
+# Verify that the POSIX 'test' builtin complains loudly when the '=~' operator is used rather than
+# failing silently. See //github.com/att/ast/issues/1152.
```
missing the scheme part of the URI should be https://github.com/att/ast/issues/1152

JohnoKing replied:
Thanks for pointing that out, now fixed.

JohnoKing committed:
Add missing https

JohnoKing committed:
Apply fix for when ]] is used as an operator

This is the patch from https://github.com/ksh93/ksh/pull/245#issuecomment-808795374.
I extended the patch with the ksh2020 error message, which
is a bit more helpful than 'unknown operator' if [[...]] happens
to support the operator 'test' rejected.

Co-authored-by: Martijn Dekker <martijn@inlv.org>

JohnoKing committed:
Add missing backslash

JohnoKing committed:
NEWS: Update with fix for ']]'

JohnoKing committed:
Quick fix for typo

JohnoKing committed:
Remove stray single quote

JohnoKing committed:
Change [[...]] -> [[ ... ]]
